### PR TITLE
[FEATURE] Indentation of legend groups and subgroups

### LINKED ISSUE #29606: Allow indenting of sub-items in a legend
Author Name: **Robin Wilson** (@robintw)
Original Redmine Issue: [21791](https://issues.qgis.org/issues/21791)

Redmine category:map_composer/printing

---
There are various spacing controls for legends in the Print Composer, but none of them seem to let you indent the subitems of a legend item. This is described in more detail, with examples, at https://gis.stackexchange.com/questions/279707/indenting-subitems-in-qgis-composer-legend - and it'd be great if this could be added.

### PR DESCRIPTION
**Pull request by jdugge**
## Description

This pull request adds two new entries to the "Spacing" section of the Legend properties:
* Indent of group items
* Indent of subgroup items

Items belonging to groups or subgroups will be shifted right by the specified amount.

Symbols on left:
image

Symbols on right:
image

Fixes #29606

### CODE REVIEW

**Comment by roya0045** [comment]
@jdugge Any plans to add the missing documentation? This seems like a quite nice feature. Are multiple columns well handled with or without the 'Even column width' option?

**Comment by nyalldawson** [comment]
I second that -- @jdugge do you have plans to revive this work? I'm waiting on all green from the CI tests before I review

**Comment by jdugge** [comment]
I'll try to fix the "Unacceptable missing documentation" error in the next couple of days.

**Comment by jdugge** [comment]
@nyalldawson Tests are passing now
@roya0045 Yes, the feature works with and without that option

**Comment by jdugge** [comment]
Is there anything that needs to be added to this PR so it can be merged?

File: src/core/qgslegendrenderer.cpp
Comment (nyalldawson) [inline src/core/qgslegendrenderer.cpp:689]: For right-to-left legend arrangements this would need to be applied to the right side of the column

File: src/core/qgslegendrenderer.cpp
Comment (nyalldawson) [inline src/core/qgslegendrenderer.cpp:683]: why are these negative values, when the similar code for drawLayerTitle uses positive values?

File: src/core/qgslegendrenderer.h
Comment (nyalldawson) [inline src/core/qgslegendrenderer.h:147]: This needs some documentation

File: src/core/qgslegendstyle.h
Comment (nyalldawson) [inline src/core/qgslegendstyle.h:126]: Can you add a note in these methods that the units are millimeters?

**Comment by nyalldawson** [comment]
Marking as freeze exempt for a week given the original submission date of this PR

File: src/core/qgslegendrenderer.cpp
Comment (jdugge) [reply src/core/qgslegendrenderer.cpp]: This is now implemented

File: src/core/qgslegendrenderer.cpp
Comment (jdugge) [reply src/core/qgslegendrenderer.cpp:685]: Negative value used here because we need to undo the indentation for the group/subgroup title (only the group items should be indented). I've added comments to the code to avoid confusion

File: src/core/qgslegendrenderer.h
Comment (jdugge) [reply src/core/qgslegendrenderer.h:151]: Docstrings added

File: src/core/qgslegendstyle.h
Comment (jdugge) [reply src/core/qgslegendstyle.h:126]: Done

**Comment by jdugge** [comment]
@nyalldawson Do I need to make any other changes to get this merged?

**Comment by nyalldawson** [comment]
@jdugge it's a bit too close to release now to push this into 3.18 -- let's delay to 3.20

**Comment by jdugge** [comment]
Is there anything left for me to do to get this merged?

**Comment by jdugge** [comment]
@nyalldawson Can this be merged or do I need to do anything else?

**Comment by roya0045** [comment]
@jdugge you got a conflict, but otherwise I think waiting is the only option.

**Comment by roya0045** [comment]
Conflict still there.

**Review by m-kuhn — APPROVED** [review]

**Comment by roya0045** [comment]
Stale-bot didn't de-stale after approval.

**Comment by jdugge** [comment]
Could this be merged so it doesn't miss another release?

**Comment by jdugge** [comment]
@m-kuhn Fantastic, thank you!

**Comment by m-kuhn** [comment]
Thank you @jdugge very nice work

**Comment by roya0045** [comment]
@jdugge Thanks Jdugge for carrying this through and @m-kuhn thanks Matthias for the merge!

**Pull request merged**
